pandas/core/frame.py

```diff
@@ -10271,18 +10271,27 @@ def quantile(
         C        1 days 12:00:00
         Name: 0.5, dtype: object
         """
+
+        data = self._get_numeric_data() if numeric_only else self
+        axis = self._get_axis_number(axis)
+        dtypes = set(data.dtypes)
+
         validate_percentile(q)
 
         if not is_list_like(q):
             # BlockManager.quantile expects listlike, so we wrap and unwrap here
             res = self.quantile(
                 [q], axis=axis, numeric_only=numeric_only, interpolation=interpolation
             )
-            return res.iloc[0]
+            res = res.iloc[0]
+            # GH#41544
+            if len(dtypes) == 1 and not numeric_only:
```

Comment:
Could the correct dtypes be set in the `if len(data.columns) == 0:` block below? It would be preferential if the bug could be address at the highest level possible.

cc @jbrockmendel if you have done some work here in the past with quantile refactoring.

Reply:
> It would be preferential if the bug could be address at the highest level possible.

do you mean lowest level possible?  i.e. if the Block method is giving an incorrect result, id want to fix it there

Reply:
tentatively looks like the problem is introduces on L10303 (L10294 in master)

Reply:
@mroeschke

If there is no data to calculate quantiles for in the target DataFrame (no columns if `axis == 0`, no rows if `axis == 1`), the code in the `if len(data.columns) == 0:` block returns a DataFrame with no columns, with the requested quantiles as the index. 

If only one quantile is asked for, the `iloc[0]` of this empty DataFrame is returned. Taking the `iloc` of a row of a DataFrame with no columns returns a Series with its dtype set to `float64`, by default. This is the cause of the observed bug.

In my view, the problem with editing the code in the `if len(data.columns) == 0:` block is that a DataFrame without any columns cannot store any dtype information (as far as I can tell, please correct if untrue).

Therefore, I see the following possible solutions:

1. Refactor the block so that instead of a DataFrame with no columns, it returns a DataFrame with one column with 'dummy' data to preserve the correct dtype.
2. Allow DataFrames with no columns to store dtype information.
3. Address the bug specifically, by setting the correct dtype in the originally identified triggering case (my initial solution).

I personally only see 1. as a good solution, but this may break the intended return 'schema' of the function, which may be unacceptable. Could you advise me regarding this?

Reply:
As mentioned in https://github.com/pandas-dev/pandas/issues/41544, if `data.T` maintained a `datetime64[ns]` dtype instead of object, would that solve this issue?

Reply:
Since `data.T` is not used to evaluate the values returned in the block, I would say no. The if statement just tests if `data` (/ `data.T`) has any columns. After that, a constructor generates an empty DataFrame as explained in my previous comment.

Also, correct me if I have misunderstood, but `data.T`, doesn't actually contain any dtypes as it is a DataFrame with no columns. It is the Series object that `.dtype` returns that has the "object" dtype:

```
>>> df = DataFrame(columns=["a", "b"], dtype="datetime64[ns]")
>>> # dtypes of a, b are datetime64[ns] and datetime64[ns] , dtype of returned series is "object"
>>> df.dtypes
a    datetime64[ns]
b    datetime64[ns]
>>> # empty list of dtypes, as df.T has no columns, though dtype of returned series is still "object"
dtype: object
>>> df.T.dtypes
Series([], dtype: object)
```

Reply:
I see.

On second thought maybe this isn't a bug per se but just defined DataFrame behavior. As you mentioned, DataFrames with no columns are designed not to store dtypes, and empty Series in pandas default to float64.